server/memcached/src/main/java/org/infinispan/server/memcached/MemcachedInboundAdapter.java

```diff
@@ -0,0 +1,217 @@
+   }
+
+   @Override
+   public void handlerAdded(ChannelHandlerContext ctx) throws Exception {
+      ctx.channel().attr(ALLOCATOR_KEY).set(size -> allocateBuffer(ctx, size));
+      progressive = MemcachedAccessLogging.isEnabled() ? ctx.newProgressivePromise() : null;
+      super.handlerAdded(ctx);
+   }
+
+   @Override
+   public void channelUnregistered(ChannelHandlerContext ctx) throws Exception {
+      if (progressive != null)
+         progressive.setFailure(new IllegalStateException("Channel unregistered"));
+      super.channelUnregistered(ctx);
+   }
+
+   @Override
+   public void channelReadComplete(ChannelHandlerContext ctx) throws Exception {
+      if (ctx.channel().config().isAutoRead()) {
+         flushBufferIfNeeded(ctx);
+      }
+      super.channelReadComplete(ctx);
+   }
+
+   @Override
+   public void channelWritabilityChanged(ChannelHandlerContext ctx) throws Exception {
+      if (resumeAutoReadOnWritability && ctx.channel().isWritable()) {
+         resumeAutoReadOnWritability = false;
+         resumeAutoRead(ctx);
+      }
+      super.channelWritabilityChanged(ctx);
+   }
+
+   @Override
+   public void channelRead(ChannelHandlerContext ctx, Object msg) {
+      if (msg == null) return;
+      handleResponse(ctx, (MemcachedResponse) msg);
+   }
+
+   @Override
+   public void exceptionCaught(ChannelHandlerContext ctx, Throwable cause) {
+      log.unexpectedException(cause);
+      if (ctx.channel().isOpen()) flushBufferIfNeeded(ctx);
+      ctx.close();
+   }
+
+   private void handleResponse(ChannelHandlerContext ctx, MemcachedResponse res) {
+      CompletionStage<?> cs = res.getResponse();
+      if (CompletionStages.isCompletedSuccessfully(cs)) {
+         Object result = CompletionStages.join(cs);
+         res.writeResponse(result, ctx.channel().attr(ALLOCATOR_KEY).get());
+         if (outbound != null && outbound.readableBytes() > ctx.channel().bytesBeforeUnwritable()) {
+            flushBufferIfNeeded(ctx, true, res);
+            ctx.channel().config().setAutoRead(false);
+            resumeAutoReadOnWritability = true;
+            return;
+         }
+         registerResponseForLater(res);
+         return;
+      }
+
+      ctx.channel().config().setAutoRead(false);
+      cs.whenCompleteAsync((obj, t) -> {
+         assert ctx.channel().eventLoop().inEventLoop();
+
+         ByteBufPool allocator = ctx.channel().attr(ALLOCATOR_KEY).get();
+         if (t != null) {
+            res.writeFailure(t, allocator);
+            flushBufferIfNeeded(ctx, false, res);
+            return;
+         }
+
+         res.writeResponse(obj, allocator);
+         flushBufferIfNeeded(ctx, false, res);
+         resumeAutoRead(ctx);
+      }, ctx.channel().eventLoop());
+   }
+
+   private void registerResponseForLater(MemcachedResponse res) {
+      if (progressive != null) {
+         progressiveEnd++;
+         progressive = progressive.addListener(new MemcachedProgressiveListener(res, progressive));
```

Comment:
I worry about this being a memory leak... were you able to confirm all of these listeners are not referencing each other? I wonder if instead we should assign progressive to be final and the listeners we register also store the progressive offset in it. Then when we progress it all the listeners that have a value <= can remove themselves and call flush?

Reply:
I updated with the approach you suggested. Analyzing from jmc and mat does not report a leak. Although, we rely on this only when the access log is enabled, and perf really takes a hit in this case.